x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -169,24 +176,36 @@ func (k msgServer) RelayPayment(goCtx context.Context, msg *types.MsgRelayPaymen
 			)
 		}
 
-		isValidPairing, allowedCU, providers, err := k.Keeper.ValidatePairingForClient(
-			ctx,
-			relay.SpecId,
-			providerAddr,
-			uint64(relay.Epoch),
-			project,
-		)
-		if err != nil {
-			return nil, utils.LavaFormatWarning("invalid pairing on proof of relay", err,
-				utils.Attribute{Key: "client", Value: clientAddr.String()},
-				utils.Attribute{Key: "provider", Value: providerAddr.String()},
-			)
-		}
-		if !isValidPairing {
-			return nil, utils.LavaFormatWarning("invalid pairing on proof of relay", fmt.Errorf("pairing result doesn't include provider"),
-				utils.Attribute{Key: "client", Value: clientAddr.String()},
-				utils.Attribute{Key: "provider", Value: providerAddr.String()},
+		// generate validate pairing cache key with CuTrackerKey() to reuse code (doesn't relate to CU tracking at all)
+		validatePairingKey := subscriptiontypes.CuTrackerKey(clientAddr.String(), relay.Provider, relay.SpecId)
+		var providers []epochstoragetypes.StakeEntry
+		allowedCU := uint64(0)
+		pairingData, ok := validatePairingCache[validatePairingKey]
+		if ok {
+			allowedCU = pairingData.allowedCU
+			providers = pairingData.providers
+		} else {
+			isValidPairing := false
+			isValidPairing, allowedCU, providers, err = k.Keeper.ValidatePairingForClient(
+				ctx,
```

Comment:
can allowed CU change in between iterations? if so we need to recalculate it

Reply:
done. see https://github.com/lavanet/lava/commit/e33ae32dd77042944191f696e87d51636ed548db